### PR TITLE
More efficient ckks encoder low-level arithmetic

### PR DESCRIPTION
**Pull request by Pro7ech**
The ckks encoding is now ~3.5x faster (omitting the final NTT step).

### CODE REVIEW

File: ckks/params.go [hunk 1]
```diff
@@ -339,15 +342,16 @@ type Parameters struct {
...
 // It returns the empty parameters Parameters{} and a non-nil error if the specified parameters are invalid.
-func NewParameters(rlweParams rlwe.Parameters, logSlot int, defaultScale float64) (p Parameters, err error) {
+func NewParameters(rlweParams rlwe.Parameters, logSlots int, defaultScale float64) (p Parameters, err error) {
 	if rlweParams.Equals(rlwe.Parameters{}) {
 		return Parameters{}, fmt.Errorf("provided RLWE parameters are invalid")
 	}
 
-	if logSlot > int(rlweParams.RingQ().NthRoot>>2) {
-		return Parameters{}, fmt.Errorf("logSlot=%d is larger than the logN-1=%d", logSlot, int(rlweParams.RingQ().NthRoot>>2))
+	if maxLogSlots := bits.Len64(rlweParams.RingQ().NthRoot) - 3; logSlots > maxLogSlots || logSlots < minLogSlots {
```
Comment (ChristianMct) [inline ckks/params.go:350]: Does this (small) change deserve a line in the CHANGELOG ? E.g., `CKKS: users can no longer instantiate parameters with less than 2^3 slots.` ?

Comment (Pro7ech) [reply ckks/params.go:350]: yes

File: ring/utils.go [hunk 2]
```diff
@@ -54,6 +54,28 @@ func ModexpMontgomery(x uint64, e int, q, qInv uint64, bredParams []uint64) (res
 	return result
 }
 
+// SliceBitReverseInPlaceFloat64 applies an in-place bit-reverse permuation on the input slice.
+func SliceBitReverseInPlaceFloat64(slice []float64, N int) {
```
Comment (Pro7ech) [inline ring/utils.go:58]: Should be un lattigo/utils package

File: ckks/params.go (hunk 1 above)
Comment (ChristianMct) [reply ckks/params.go:350]: Done

File: ring/utils.go (hunk 2 above)
Comment (ChristianMct) [reply ring/utils.go]: Done

**Review by ChristianMct — APPROVED** [review]

**Review by jrtroncoso — CHANGES_REQUESTED** [review]
bfv/bfv_vector_ops.go and ckks/ckks_benchmarks_test.go require updates on the imports

**Pull request merged**